mail_tracking_mailgun/models/mail_tracking_email.py

```diff
@@ -0,0 +1,198 @@
+        if mailgun_event_type not in self._mailgun_supported_event_types:
+            _logger.info("Mailgun: event type '%s' not supported",
+                         mailgun_event_type)
+            return False
+        # OK, event type is valid
+        return True
+
+    def _mailgun_signature(self, api_key, timestamp, token):
+        return hmac.new(
+            key=str(api_key),
+            msg='{}{}'.format(str(timestamp), str(token)),
+            digestmod=hashlib.sha256).hexdigest()
+
+    def _mailgun_signature_verify(self, event):
+        event = event or {}
+        api_key = self.env['ir.config_parameter'].get_param('mailgun.apikey')
+        if not api_key:
+            _logger.info("No Mailgun api key configured. "
+                         "Please add 'mailgun.apikey' to System parameters "
+                         "to enable Mailgun authentication webhoook requests. "
+                         "More info at: "
+                         "https://documentation.mailgun.com/user_manual.html"
+                         "#webhooks")
+        else:
+            timestamp = event.get('timestamp')
+            token = event.get('token')
+            signature = event.get('signature')
+            event_digest = self._mailgun_signature(api_key, timestamp, token)
+            if signature != event_digest:
+                _logger.error("Mailgun: Invalid signature '%s' != '%s'",
+                              signature, event_digest)
+                return False
+        # OK, signature is valid
+        return True
+
+    def _db_verify(self, event):
+        event = event or {}
+        odoo_db = event.get('odoo_db')
+        current_db = self.env.cr.dbname
+        if odoo_db != current_db:
+            _logger.info("Mailgun: Database '%s' is not the current database",
+                         odoo_db)
+            return False
+        # OK, DB is current
+        return True
+
+    def _mailgun_metadata(self, mailgun_event_type, event, metadata):
+        # Get Mailgun timestamp when found
+        ts = event.get('timestamp', False)
+        try:
+            ts = float(ts)
+        except:
+            ts = False
+        if ts:
+            dt = datetime.utcfromtimestamp(ts)
+            metadata.update({
+                'timestamp': ts,
+                'time': fields.Datetime.to_string(dt),
+                'date': fields.Date.to_string(dt),
+            })
+        # Common field mapping
+        mapping = {
+            'recipient': 'recipient',
+            'ip': 'ip',
+            'user_agent': 'user-agent',
+            'os_family': 'client-os',
+            'ua_family': 'client-name',
+            'ua_type': 'client-type',
+            'url': 'url',
+        }
+        for k, v in mapping.iteritems():
+            if event.get(v, False):
+                metadata[k] = event[v]
+        # Special field mapping
+        metadata.update({
+            'mobile': event.get('device-type') in ('mobile', 'tablet'),
```

Comment:
Of course for 2 elements this becomes almost irrelevant, but just wanted to share why is it good to use sets when all you want is to use `in`: http://stackoverflow.com/a/17945009/1468388 :wink: